### ISSUE TITLE
example code for swimmer environment is not reaching an optimal policy

### ISSUE THREAD

**Issue by kishanpb** [rlworkgroup/garage#1005]
This is final epoch values from trpo_swimmer_ray_sampler.py code:
---------------------------------------  ------------
AverageDiscountedReturn                  21.5863
AverageReturn                            40.3846
Entropy                                   2.61494
Extras/EpisodeRewardMean                 40.2554
GaussianMLPPolicy/Entropy                 2.61755
GaussianMLPPolicy/KL                      0.00660417
GaussianMLPPolicy/KLBefore                0
GaussianMLPPolicy/LossAfter              -0.0175566
GaussianMLPPolicy/LossBefore              2.28882e-08
GaussianMLPPolicy/dLoss                   0.0175566
Iteration                                39
LinearFeatureBaseline/ExplainedVariance   0.745997
MaxReturn                                44.2927
MinReturn                                35.879
NumTrajs                                  8
Perplexity                               13.6664
StdReturn                                 2.60611
---------------------------------------  ------------
and this is from trpo_swimmer.py
---------------------------------------  ------------
AverageDiscountedReturn                  23.5388
AverageReturn                            30.4157
Entropy                                   2.72082
EnvExecTime                               0.371212
Extras/EpisodeRewardMean                 28.8893
GaussianMLPPolicy/Entropy                 2.71873
GaussianMLPPolicy/KL                      0.00910066
GaussianMLPPolicy/KLBefore                0
GaussianMLPPolicy/LossAfter              -0.0217502
GaussianMLPPolicy/LossBefore             -8.58307e-09
GaussianMLPPolicy/dLoss                   0.0217502
Iteration                                39
LinearFeatureBaseline/ExplainedVariance   0.989434
MaxReturn                                32.676
MinReturn                                27.2979
NumTrajs                                  8
Perplexity                               15.1928
PolicyExecTime                            0.409919
ProcessExecTime                           0.0361636
StdReturn                                 1.85247
---------------------------------------  ------------

In both cases, the averageReturn is just around 30, whereas swimmer environment optimal max reward is 360. When I see how the policy is doing using the sim_policy.py script, the 3 link swimmer is just curling in, and not moving forward at all! According to the paper https://arxiv.org/pdf/1604.06778.pdf, rllab (a historical version of garage) should reach an average reward of 96 in trpo.

I am observing a similar thing with vpg on swimmer as well. It's reaching 30 but not 92.3 as per the paper.

Has anyone improved on the current examples and got a better policy?

**Comment by ryanjulian** [comment]
@kishanpb thanks for reporting an issue.

We takes performance very seriously. With that said, the examples are meant to be illustrations of how to use the toolkit, not fully-tuned solutions with optimal hyperparameters.

There are many reason you should expect to see a discrepancy from Duan, et al.:
* This example runs 40 optimization epochs with a batch size of 4000, for a total of 160,000 environment steps. Duan, et al. used 500 optimization epochs with a batch size of 50k (see Appendix 2), for a total of 25,000,000 environment steps. That's a huge difference.
* The version of Swimmer from Duan, et al. is different than the one used in the example. Duan et al. used a version of `Swimmer` which predates Open AI Gym (the experiments from that paper became the foundation of gym, but gym rewrote them significantly). To make a comparison, you would have to run your benchmark on the original version [here](https://github.com/rll/rllab/blob/master/rllab/envs/mujoco/swimmer_env.py)
* This example uses a different learning rate than the one reported in the paper
* There are [reports of a regression](https://github.com/hill-a/stable-baselines/issues/500) in the trainability of `Swimmer-v2`

If you find a set of hyperparameters which solve `Swimmer-v2`, please feel free to send a PR. We'd love to update the code with tuned hyperparamters.

I'm going to close this issue for now. Please feel free to open another one if you find a breakage.

**Comment by kishanpb** [comment]
Okay, then what does the benchmark folder contain? Even they aren't tuned to reach optimality? Although I'll try tuning as well.

**Comment by ryanjulian** [comment]
`tests/benchmarks` contains comparative benchmark scripts which allow developers of the toolkit to ensure that their code changes do not compromise the performance of algorithms or algorithm components (we call those components primitives).

They do not contain tuned examples of optimal hyperparameters.

Typically these scripts are designed to run the same algorithm before and after a code change, or to compare the performance of a garage algorithm with an algorithm from a different library, using exactly the same hyperparameters and random seeds.

Their purpose is not to ensure that the library achieves some **fixed performance** `p` on environment `E` using algorithm `A` with hyperparameters `h`, but to ensure that algorithm version `A'` achieves **the same or better performance** `p` as algorithm version `A` on environment `E` using hyperparameters `h`.

**Comment by kishanpb** [comment]
@ryanjulian, thanks for the detailed explanation! Is it possible for you to include these in the docs folder? The present documentation is too minimal!

I was looking at the examples folder, just like you've provided a sim_policy.py, can you also add a reward plot script? Or, is it already available which I missed locating?

**Comment by ryanjulian** [comment]
@kishanpb we would love to have more documentation and examples.

As an open source project, we are reliant on great work and contributions from people like you to stay alive. **If you see something in the library you'd like to see improved, please feel free to add it and send a PR.** You can check out [CONTRIBUTING.md](https://github.com/rlworkgroup/garage/blob/master/CONTRIBUTING.md) for details on how to prepare a PR for garage.

Information like what I shared above is usually considered developer documentation, and often gets the lowest priority in projects like this (because the developers can just share it amongst themselves).

For plotting your algorithm's performance, I recommend launching TensorBoard against directory `data` in whichever directory you launched your experiment from. That's where garage will put log files, including tfevents files readable by TensorBoard.

i.e.

```
cd path/to/foo
python foo_experiment.py
# open a new terminal
cd path/to/foo
tensorboard --logdir=data/local
```

If you are working out of a copy of this github repo, `path/to/foo` is probably the garage git respository.

**Comment by kishanpb** [comment]
Okay, I'll contribute once I am successful at implementing these. Thanks for the suggestion.

So, there's a progress.csv log file in the data folder which contains all the required data. I can write a processing script to plot the performance curves. But, what is the input to `tensorboard`?  Is it the path to this progress.csv file?

**Comment by ryanjulian** [comment]
It is the directory, exactly as I showed in the example.

You might want to take a look at garage.experiment.nb_utils for tools which help process progress.csv, thought those have not been tried in a long time and are untested.

You can also try using viskit if you'd like, thought it's not very heavily maintained. I recommend you stick to TensorBoard for now.

**Comment by kishanpb** [comment]
Okay, I'll explore. I asked this because when I ran this `tensorboard --logdir=data/local`, no plot prompted. Also, local directory consists of many experiments which I have already run, but I don't want to process every experiment which I have run until now!

**Comment by ryanjulian** [comment]
You open your web browser to `localhost:6006` to view TensorBoard. Please read the output on the terminal and also search for some examples if you need information on how to use TensorBoard.
